In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv("AviationData.csv", encoding='latin1', low_memory=False)


In [3]:
df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

The AviationData dataset has 31 columns and 88889 rows or observations, which have both float and character data types. Next, a quick peek into the summary statisics of the numeric variables present in the dataset.

df.describe()

What about the column names? 

In [6]:
df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

Since the column names  have a uniform naming system there are no changes needed. Let's proceed to create a copy. 

In [7]:
df1= df.copy(deep=True)

In [9]:
df1.isna().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38640
Airport.Name              36099
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1317
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7077
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6381
Publication.Date          13771
dtype: i

With the dataset having 88889 observations we have to proceed to remove columns with more than half of the obersavtions missing. Additionally, any column deemed not statistically significant.

In [10]:
half_threshold = len(df) / 2
cols_with_half_missing = df.columns[df.isnull().sum() > half_threshold]

# Print or list them
print("Columns with more than 50% missing values:")
print(cols_with_half_missing)

Columns with more than 50% missing values:
Index(['Latitude', 'Longitude', 'Aircraft.Category', 'FAR.Description',
       'Schedule', 'Air.carrier'],
      dtype='object')


In [13]:
df1 = df1.dropna(thresh=half_threshold, axis=1)
df1.isna().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Airport.Code              38640
Airport.Name              36099
Injury.Severity            1000
Aircraft.damage            3194
Registration.Number        1317
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7077
Purpose.of.flight          6192
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6381
Publication.Date          13771
dtype: int64

In [14]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 25 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Airport.Code            50249 non-null  object 
 7   Airport.Name            52790 non-null  object 
 8   Injury.Severity         87889 non-null  object 
 9   Aircraft.damage         85695 non-null  object 
 10  Registration.Number     87572 non-null  object 
 11  Make                    88826 non-null  object 
 12  Model                   88797 non-null  object 
 13  Amateur.Built           88787 non-null  object 
 14  Number.of.Engines       82805 non-null

In [16]:
for col in df1.columns:
    unique_vals = df1[col].unique()
    print(f"{col} — {len(unique_vals)} unique values")
    print(unique_vals[:10])  # Show first 10 unique values
    print("-" * 40)


Event.Id — 87951 unique values
['20001218X45444' '20001218X45447' '20061025X01555' '20001218X45448'
 '20041105X01764' '20170710X52551' '20001218X45446' '20020909X01562'
 '20020909X01561' '20020909X01560']
----------------------------------------
Investigation.Type — 2 unique values
['Accident' 'Incident']
----------------------------------------
Accident.Number — 88863 unique values
['SEA87LA080' 'LAX94LA336' 'NYC07LA005' 'LAX96LA321' 'CHI79FA064'
 'NYC79AA106' 'CHI81LA106' 'SEA82DA022' 'NYC82DA015' 'MIA82DA029']
----------------------------------------
Event.Date — 14782 unique values
['1948-10-24' '1962-07-19' '1974-08-30' '1977-06-19' '1979-08-02'
 '1979-09-17' '1981-08-01' '1982-01-01' '1982-01-02' '1982-01-03']
----------------------------------------
Location — 27759 unique values
['MOOSE CREEK, ID' 'BRIDGEPORT, CA' 'Saltville, VA' 'EUREKA, CA'
 'Canton, OH' 'BOSTON, MA' 'COTTON, MN' 'PULLMAN, WA' 'EAST HANOVER, NJ'
 'JACKSONVILLE, FL']
----------------------------------------
Co

The above steps check the unique values in each of the remaining columns. Given the large size of this dataset, it is safe to assume that columns whose unique values caount is the close to or same as the total observations can be considered as not relevant for any analysis. This is because they do bring much value or rather have nothing to offer research wise. These columns are Event.Id, Accident.Number, and Registration.Number.

In [17]:
columns_to_drop = ['Event.Id', 'Accident.Number', 'Registration.Number']
df1 = df1.drop(columns=columns_to_drop, )


In [18]:
df1.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Columns: 22 entries, Investigation.Type to Publication.Date
dtypes: float64(5), object(17)
memory usage: 14.9+ MB


In [ ]:
Now let's deal with the missing values in the object dtype columns by replacing the missing values with "unknown'. This was preferred instead of mode since the missing values are many and cause to much skeweness.  

In [19]:
for col in df1.select_dtypes(include=['object']).columns:
    df1[col] = df1[col].fillna('Unknown')


In [ ]:
For the numeric dtypes we can replace the unkowns with mean

In [21]:
for col in df1.select_dtypes(include='float').columns:
    if df1[col].isnull().any():
        mean_val = df1[col].mean()
        df1[col] = df1[col].fillna(mean_val)

In [22]:
df1.isna().sum()

Investigation.Type        0
Event.Date                0
Location                  0
Country                   0
Airport.Code              0
Airport.Name              0
Injury.Severity           0
Aircraft.damage           0
Make                      0
Model                     0
Amateur.Built             0
Number.of.Engines         0
Engine.Type               0
Purpose.of.flight         0
Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
Weather.Condition         0
Broad.phase.of.flight     0
Report.Status             0
Publication.Date          0
dtype: int64

In [23]:
assert type(df) == pd.DataFrame

In [24]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 22 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Investigation.Type      88889 non-null  object 
 1   Event.Date              88889 non-null  object 
 2   Location                88889 non-null  object 
 3   Country                 88889 non-null  object 
 4   Airport.Code            88889 non-null  object 
 5   Airport.Name            88889 non-null  object 
 6   Injury.Severity         88889 non-null  object 
 7   Aircraft.damage         88889 non-null  object 
 8   Make                    88889 non-null  object 
 9   Model                   88889 non-null  object 
 10  Amateur.Built           88889 non-null  object 
 11  Number.of.Engines       88889 non-null  float64
 12  Engine.Type             88889 non-null  object 
 13  Purpose.of.flight       88889 non-null  object 
 14  Total.Fatal.Injuries    88889 non-null

In [ ]:
# Exploratort Data Analysis

In [30]:
# Top 10 Locations
print("Top 10 Locations:")
print(df1['Location'].value_counts().head(10).to_string()+"\n")

# Top 10 Countries
print("Top 10 Countries:")
print(df1['Country'].value_counts().head(10).to_string()+"\n")


# Top 10 Makes
print("Top 10 Makes:")
print(df1['Make'].value_counts().head(10).to_string()+"\n")


# Top 10 Models
print("Top 10 Models:")
print(df1['Model'].value_counts().head(10).to_string()+"\n")


Top 10 Locations:
ANCHORAGE, AK      434
MIAMI, FL          200
ALBUQUERQUE, NM    196
HOUSTON, TX        193
CHICAGO, IL        184
FAIRBANKS, AK      174
TUCSON, AZ         142
ORLANDO, FL        132
PHOENIX, AZ        132
ENGLEWOOD, CO      131

Top 10 Countries:
United States     82248
Brazil              374
Canada              359
Mexico              358
United Kingdom      344
Australia           300
France              236
Unknown             229
Spain               226
Bahamas             216

Top 10 Makes:
Cessna     22227
Piper      12029
CESSNA      4922
Beech       4330
PIPER       2841
Bell        2134
Boeing      1594
BOEING      1151
Grumman     1094
Mooney      1092

Top 10 Models:
152          2367
172          1756
172N         1164
PA-28-140     932
150           829
172M          798
172P          689
182           659
180           622
150M          585

